### ISSUE TITLE
Auto capitalization on dynamic keyboards

### ISSUE THREAD

**Issue by annakirkpatrick** [OptiKey/OptiKey#477]
As referenced in [this issue](https://github.com/OptiKey/OptiKey/issues/471), I have a custom dynamic keyboard up and running (XML source for that keyboard [here](https://github.com/OptiKey/OptiKey/blob/master/src/JuliusSweetland.OptiKey/Resources/DynamicKeyboards/BigHackerKeyboard.xml)). It's awesome, but it seems I'm now finding all the little bugs. :-)

With the default keyboards, OptiKey's default behavior is to automatically press the shift key after the user types a period followed by a space. (This behavior occurs with other end of sentence punctuation, too.)

On my dynamic keyboard, after typing a period and then a space, I can see the shift key briefly change color (indicating the key is pressed down) but then it very quickly changes back to the default white color (indicating the key is now up). The net result is no auto capitalization after the end of a sentence.

I think what is actually happening is that the auto capitalization is being suppressed as "probably unhelpful." When I uncheck the option "… But intelligently suppress auto-caps where it is probably unhelpful," I get the correct auto capitalization behavior after periods.

Since this case is pretty clearly helpful auto-caps behavior, I think this is a bug in whatever code is being used to determine when auto-caps is probably unhelpful.

I am attaching a log file.  The auto-caps options were set to their default states: auto-caps on and suppression where unhelpful also on. I typed "This is a test. " and saw that auto caps did not correctly occur after my  period and space. (Actually, that was what I meant to type. I actually typed "This[is a test. " The behavior is the same.)

The whole thing is attached below, but this appears to be the relevant part: (this quote starts on line 60678)
```
2018-07-12 12:10:45,359 [ 1] DEBUG JuliusSweetland.OptiKey.Observables.TriggerSources.KeyboardKeyDownUpSource: Trigger key down detected (F13)
2018-07-12 12:10:45,360 [ 1] DEBUG JuliusSweetland.OptiKey.Services.InputService: Selection trigger signal (with relevent PointAndKeyValue) detected.
2018-07-12 12:10:45,360 [ 1] DEBUG JuliusSweetland.OptiKey.Services.InputService: Selection mode is KEY and the key on which the trigger occurred is enabled.
2018-07-12 12:10:45,360 [ 1] DEBUG JuliusSweetland.OptiKey.Services.InputService: Publishing Selection event with PointAndKeyValue:(1141.93398268329,650.337313058449)='[Char: |Unicode:U+0020]'
2018-07-12 12:10:45,360 [ 1] INFO  JuliusSweetland.OptiKey.UI.ViewModels.MainViewModel: Selection event received from InputService.
2018-07-12 12:10:45,360 [ 1] INFO  JuliusSweetland.OptiKey.Services.AudioService: Playing sound 'Resources\Sounds\Click1.wav' at volume '50'
2018-07-12 12:10:45,363 [ 1] INFO  JuliusSweetland.OptiKey.UI.ViewModels.MainViewModel: Firing KeySelection event with KeyValue '[Char: |Unicode:U+0020]'
2018-07-12 12:10:45,363 [ 1] DEBUG JuliusSweetland.OptiKey.Services.InputService: Publishing Selection Result event with 1 point(s), FunctionKey:'', String:'[Char: |Unicode:U+0020]', Best match '', Suggestion count:
2018-07-12 12:10:45,363 [ 1] INFO  JuliusSweetland.OptiKey.UI.ViewModels.MainViewModel: SelectionResult event received from InputService.
2018-07-12 12:10:45,363 [ 1] INFO  JuliusSweetland.OptiKey.UI.ViewModels.MainViewModel: KeySelectionResult received with string value '[Char: |Unicode:U+0020]' and function key values ''
2018-07-12 12:10:45,363 [ 1] DEBUG JuliusSweetland.OptiKey.Services.KeyboardOutputService: Processing single key captured text '[Char: |Unicode:U+0020]'
2018-07-12 12:10:45,363 [ 1] INFO  JuliusSweetland.OptiKey.Services.KeyboardOutputService: Combing dead keys (diacritics) on ' '
2018-07-12 12:10:45,363 [ 1] DEBUG JuliusSweetland.OptiKey.Services.KeyboardOutputService: Processing captured text '[Char: |Unicode:U+0020]'
2018-07-12 12:10:45,363 [ 1] DEBUG JuliusSweetland.OptiKey.Services.KeyboardOutputService: Suppressing auto space before this capture as the user appears to be typing one char at a time. Also the last text change was not a suggestion, and the current capture is not a single letter captured with multi-key selection enabled.
2018-07-12 12:10:45,363 [ 1] DEBUG JuliusSweetland.OptiKey.Services.KeyboardOutputService: Storing last text change ' '
2018-07-12 12:10:45,363 [ 1] INFO  JuliusSweetland.OptiKey.Services.KeyboardOutputService: Publishing '[Char: |Unicode:U+0020]' => as virtual key code SPACE (using hard coded mapping)
2018-07-12 12:10:45,363 [ 1] DEBUG JuliusSweetland.OptiKey.Services.PublishService: Simulating key press (down & up): SPACE
2018-07-12 12:10:45,363 [ 1] DEBUG JuliusSweetland.OptiKey.Observables.TriggerSources.KeyboardKeyDownUpSource: Key down detected with KeyCode:Space | KeyValue:32 | KeyData:Space | Modifiers:None | Alt:False | Control:False | Shift:False
2018-07-12 12:10:45,364 [ 1] DEBUG JuliusSweetland.OptiKey.Observables.TriggerSources.KeyboardKeyDownUpSource: Key up detected with KeyCode:F13 | KeyValue:124 | KeyData:F13 | Modifiers:None | Alt:False | Control:False | Shift:False
2018-07-12 12:10:45,364 [ 1] DEBUG JuliusSweetland.OptiKey.Observables.TriggerSources.KeyboardKeyDownUpSource: Trigger key up detected (F13)
2018-07-12 12:10:45,365 [ 1] DEBUG JuliusSweetland.OptiKey.Observables.TriggerSources.KeyboardKeyDownUpSource: Key up detected with KeyCode:Space | KeyValue:32 | KeyData:Space | Modifiers:None | Alt:False | Control:False | Shift:False
2018-07-12 12:10:45,366 [ 1] INFO  JuliusSweetland.OptiKey.Services.KeyboardOutputService: ReleaseUnlockedKeys called.
2018-07-12 12:10:45,367 [ 1] DEBUG JuliusSweetland.OptiKey.Services.KeyboardOutputService: Auto-pressing shift.
2018-07-12 12:10:45,378 [ 1] DEBUG JuliusSweetland.OptiKey.Services.KeyboardOutputService: Generating auto complete suggestions from 'This[is a test. '.
2018-07-12 12:10:45,378 [ 1] DEBUG JuliusSweetland.OptiKey.Services.Suggestions.NGramAutoComplete: GetSuggestions called with root ''.
2018-07-12 12:10:45,378 [ 1] DEBUG JuliusSweetland.OptiKey.Services.KeyboardOutputService: 0 suggestions generated (possibly capped to 20 by MaxDictionaryMatchesOrSuggestions setting)
2018-07-12 12:10:45,388 [ 1] DEBUG JuliusSweetland.OptiKey.Services.PublishService: Simulating key down LSHIFT
2018-07-12 12:10:45,388 [ 1] DEBUG JuliusSweetland.OptiKey.Observables.TriggerSources.KeyboardKeyDownUpSource: Key down detected with KeyCode:LShiftKey | KeyValue:160 | KeyData:LShiftKey | Modifiers:None | Alt:False | Control:False | Shift:False
2018-07-12 12:10:45,399 [ 1] DEBUG JuliusSweetland.OptiKey.Services.PublishService: Simulating key up: LSHIFT
2018-07-12 12:10:45,399 [ 1] DEBUG JuliusSweetland.OptiKey.Observables.TriggerSources.KeyboardKeyDownUpSource: Key up detected with KeyCode:LShiftKey | KeyValue:160 | KeyData:LShiftKey, Shift | Modifiers:Shift | Alt:False | Control:False | Shift:True
```

[OptiKey.log](https://github.com/OptiKey/OptiKey/files/2189422/OptiKey.log)

**Comment by mbise1993** [comment]
Hi! I discovered this project through Hacktoberfest and I'd love to make some contributions. After looking through the code a bit, this seems like a good issue to start with. I was able to reproduce the problem and have a really basic fix working on my local copy by adding a call to the `KeyboardView` base class constructor from here: https://github.com/OptiKey/OptiKey/blob/86a871f50e8ba79c6f5b1a62fe0f83519a7dce0c/src/JuliusSweetland.OptiKey/UI/Views/Keyboards/Common/DynamicKeyboard.xaml.cs#L25
which simply passes in `shiftAware: true`. However, I realize this might not be optimal for ALL dynamic keyboards and it might make sense to add an option to the dynamic keyboard XML for this. @JuliusSweetland I'd like to get your input before submitting a PR to determine which route I should take.
